stdlib/Random/src/RNGs.jl

```diff
@@ -19,23 +19,16 @@ if Sys.iswindows()
     end
 else # !windows
     struct RandomDevice <: AbstractRNG
-        file::IOStream
         unlimited::Bool
 
-        RandomDevice(; unlimited::Bool=true) =
-            new(open(unlimited ? "/dev/urandom" : "/dev/random"), unlimited)
+        RandomDevice(; unlimited::Bool=true) = new(unlimited)
     end
 
-    rand(rd::RandomDevice, sp::SamplerBoolBitInteger) = read( rd.file, sp[])
+    rand(rd::RandomDevice, sp::SamplerBoolBitInteger) = read(getfile(rd), sp[])
 
-    function serialize(s::AbstractSerializer, rd::RandomDevice)
-        Serialization.serialize_type(s, typeof(rd))
-        serialize(s, rd.unlimited)
-    end
-    function deserialize(s::AbstractSerializer, t::Type{RandomDevice})
-        unlimited = deserialize(s)
-        return RandomDevice(unlimited=unlimited)
-    end
+    getfile(rd::RandomDevice) = @inbounds DEV_RANDOM[1 + rd.unlimited]
```

Comment:
This should not be `@inbounds` as it is not essential. And while equivalent, I think using `?:` will read more clearly.

I'd also suggest going with the lazy approach you mentioned.

```suggestion
    function getfile(rd::RandomDevice)
        dev = rd.unlimited ? 2 : 1
        isassigned(DEV_RANDOM, dev) || (DEV_RANDOM[dev] = open(rd.unlimited ? "/dev/urandom" : "/dev/urandom"))
        return DEV_RANDOM[dev]
    end
```
```julia
function __init__()
    ...
    Sys.iswindows() || resize!(empty!(DEV_RANDOM), 2)
    nothing
 end
```

Reply:
> I'd also suggest going with the lazy approach you mentioned.

Your suggestion is basically what I had in the [second commit](https://github.com/JuliaLang/julia/pull/27936/commits/b1b0471655b101855ad6f3e5c00fc5d8a74b7244#diff-974d92cfd230b3e9a0dda204d56a8c9cR29), but @chethega [mentioned](https://github.com/JuliaLang/julia/pull/27936#issuecomment-548620176) that this looks thread unsafe:

> There is a data race on initialization of the RNG, but that's benign: worst case, we open nthreads many file descriptors to /dev/urandom.
> https://github.com/JuliaLang/julia/pull/27936#issuecomment-548620176

I was quite conviced. That's why I then gave up laziness because I found it overkill to have one file per thread ([third commit](https://github.com/JuliaLang/julia/pull/27936/commits/ce9069d2fbf63b6599e049ad02a403dba5f5a8fb)), and annoying to explicitly use locks.

But I'm not well versed in data-races, what is the worst case here? to open a file twice? is `DEV_RANDOM[dev]` in your example always pointing to a valid file even if the location has been written concurrently? 

(and thanks for your review!)

Reply:
Yeah, you'd potentially leak up to `nthreads` copies of the file descriptors instead of 2. We _really_ need to add a [`Threads.once`](https://linux.die.net/man/3/pthread_once) function soon, since it's rather absurd to be making design decisions based on the lack of a simple little helper functions.

Reply:
Oh, I had misunderstood @chethega's comment, which I thought was suggesting to "open nthread many file descriptors" ourselves, i.e. in an array like the `default_rng()`, to avoid the data-race! Sorry for the noise.
So I updated accordingly, with a TODO to add "thread-once" when availble.